### PR TITLE
fix: added sentry exception for invalid DynamicBindingListPath

### LINKED ISSUE #10712: [Task]: Check and throw error if there is a discrepancy in dynamicBindingPathList before saving
### Is there an existing issue for this?

- [X] I have searched the existing issues

### SubTasks
Related to https://github.com/appsmithorg/appsmith/issues/10037

We are seeing a lot of sentry errors lately related to invalid bindingPathList across multiple widgets(refer below). This error is being thrown from the server, hence we don't have much context on how we ended up with an invalid list. As part of this task, we need to add a check, before we make a save call, to identify this error on the client and then throw a sentry error. That way will have more context on the issue.
<pre>
 "widgetType" : "LIST_WIDGET",  "bindingPath" : "template.ButtonGroup1.groupButtons.groupButtonir5dikkud2.label",  "message" : "Binding path in the widget not found. Please reach out to Appsmith customer support to resolve this.",  "widgetName" : "List1",  "widgetId" : "spuvlxlf5d",  "pageId" : "61ededd4be698f35db550072",  "layoutId" : "61ededd4be698f35db550070",  "dynamicBinding" : "SWOT"
<pre>

### PR DESCRIPTION
**Pull request by keyurparalkar**
## Description
**Summary of changes**

- Added a function called `isInvalidDynamicBindingPath` in the PageSagas.tsx

**Motivation**

- This function will log error into the sentry whenever the property value is not Dynamic and the bindingPath is still present in the `dynamicBindingPathList`
- Thus the function - Checks if dynamicBindingPathList contains a property path that doesn't have a binding

Fixes #10712

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

I have update the unit test file: `helpers.test.ts` 
The test case checks the following:

- Check if the DSL is not altered
- Checks if dynamicBindingPathList contains a property path that doesn't have a binding

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/fix-log_sentry_error_dynamic_binding_path 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.06 **(0)** | 36.62 **(0)** | 34.79 **(0.01)** | 55.46 **(0)**
 :green_circle: | app/client/src/sagas/PageSagas.tsx | 24.42 **(0.22)** | 5.88 **(0)** | 14.71 **(0)** | 26.19 **(0.25)**
 :red_circle: | app/client/src/utils/WorkerUtil.ts | 89.76 **(0)** | 70.59 **(-1.96)** | 100 **(0)** | 93.33 **(0)**
 :green_circle: | app/client/src/utils/helpers.tsx | 67.09 **(1.31)** | 34.57 **(2.35)** | 60.38 **(1.56)** | 65.31 **(1.63)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**</details>

### CODE REVIEW

File: app/client/src/utils/helpers.test.ts [hunk 1]
```diff
@@ -187,3 +194,355 @@ describe("#getLocale", () => {
     expect(getLocale()).toBe(navigator.languages[0]);
   });
 });
+
+describe("#isInvalidDynamicBindingPath", () => {
+  it("DSL should not alter", () => {
```
Comment (sbalaji1192) [inline app/client/src/utils/helpers.test.ts:199]: DSL should not be altered

File: app/client/src/utils/helpers.tsx [hunk 2]
```diff
@@ -662,3 +666,50 @@ export const mergeWidgetConfig = (target: any, source: any) => {
...
+export const isInvalidDynamicBindingPath = (
+  currentDSL: Readonly<DSLWidget>,
+) => {
+  //Get the propertyPaneConfig for the current DSL Type
+  const propertyPaneConfig = WidgetFactory.getWidgetPropertyPaneConfig(
+    currentDSL.type,
+  );
+
+  //Get the bindingPaths from the dsl and propertyPaneConfig
+  const { bindingPaths } = getAllPathsFromPropertyConfig(
```
Comment (sbalaji1192) [inline app/client/src/utils/helpers.tsx:684]: We don't need to check all the bindingPaths in a widget. We just to need to check the keys in the `dynamicBindingPathList`

**Review by sbalaji1192 — CHANGES_REQUESTED** [review]

File: app/client/src/sagas/PageSagas.tsx [hunk 3]
```diff
@@ -378,6 +378,12 @@ function* savePageSaga(action: ReduxAction<{ isRetry?: boolean }>) {
       payload: savePageRequest.dsl,
     });
 
+    isInvalidDynamicBindingPath(
```
[the remark this answers is not in the document]
Comment (keyurparalkar) [reply app/client/src/sagas/PageSagas.tsx]: Changed the name to `captureInvalidDynamicBindingPath`

File: app/client/src/utils/helpers.test.ts (hunk 1 above)
Comment (keyurparalkar) [reply app/client/src/utils/helpers.test.ts]: Done

File: app/client/src/utils/helpers.tsx (hunk 2 above)
Comment (keyurparalkar) [reply app/client/src/utils/helpers.tsx:684]: Changed the logic to traverse through `dynamicBindingPathList` instead of `bindingPaths`

**Review by sbalaji1192 — APPROVED** [review]

File: app/client/src/sagas/PageSagas.tsx [hunk 4]
```diff
@@ -378,6 +378,12 @@ function* savePageSaga(action: ReduxAction<{ isRetry?: boolean }>) {
       payload: savePageRequest.dsl,
     });
 
+    captureInvalidDynamicBindingPath(
```
Comment (riodeuno) [inline app/client/src/sagas/PageSagas.tsx:381]: Please check `migrateIncorrectDynamicBindingPathLists` on line 451. I believe it is doing the same thing there. Can we re-use it?

Comment (keyurparalkar) [reply app/client/src/sagas/PageSagas.tsx:381]: @riodeuno We needed to execute a `captureInvalidDynamicBindingPath` before saving a page. And it seems the `migrateIncorrectDynamicBindingPathLists` might only get executed in the 2nd try. Capturing the error before from the client end might be beneficial.

@sbalaji1192 can you also provide your comments on this.

Comment (riodeuno) [reply app/client/src/sagas/PageSagas.tsx:381]: I was only wondering if we can re-use the `migrateIncorrectDynamicBindingPathLists` function in our case. Is that possible?

Comment (riodeuno) [reply app/client/src/sagas/PageSagas.tsx:381]: Whoops! Please ignore my comments, I completely missed the point of this PR. We're creating new sentry exceptions, not fixing the issues. My bad. 😓

**Review by riodeuno — APPROVED** [review]

File: app/client/src/sagas/PageSagas.tsx (hunk 4 above)
Comment (keyurparalkar) [reply app/client/src/sagas/PageSagas.tsx:381]: No problem @riodeuno

**Pull request merged**